// mat-file-upload/bundles/mat-file-upload.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/button'), require('@angular/common'), require('@angular/material/icon'), require('@angular/forms'), require('@angular/material/input'), require('@angular/material/select')) :
    typeof define === 'function' && define.amd ? define('mat-file-upload', ['exports', '@angular/core', '@angular/material/button', '@angular/common', '@angular/material/icon', '@angular/forms', '@angular/material/input', '@angular/material/select'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['mat-file-upload'] = {}, global.ng.core, global.ng.material.button, global.ng.common, global.ng.material.icon, global.ng.forms, global.ng.material.input, global.ng.material.select));
}(this, (function (exports, i0, i1, i2, i3, forms, input, select) { 'use strict';

    var MatFileUploadService = /** @class */ (function () {
        function MatFileUploadService() {
        }
        return MatFileUploadService;
    }());
    MatFileUploadService.ɵfac = function MatFileUploadService_Factory(t) { return new (t || MatFileUploadService)(); };
    MatFileUploadService.ɵprov = i0.ɵɵdefineInjectable({ token: MatFileUploadService, factory: MatFileUploadService.ɵfac, providedIn: 'root' });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(MatFileUploadService, [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root'
                    }]
            }], function () { return []; }, null);
    })();

    var _c0 = ["fileInput"];
    function MatFileUploadComponent_button_7_Template(rf, ctx) {
        if (rf & 1) {
            var _r4_1 = i0.ɵɵgetCurrentView();
            i0.ɵɵelementStart(0, "button", 6);
            i0.ɵɵlistener("click", function MatFileUploadComponent_button_7_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r4_1); var ctx_r3 = i0.ɵɵnextContext(); return ctx_r3.uploadFiles(); });
            i0.ɵɵtext(1);
            i0.ɵɵelementEnd();
        }
        if (rf & 2) {
            var ctx_r1 = i0.ɵɵnextContext();
            i0.ɵɵproperty("type", ctx_r1.uploadButtonType)("disabled", !ctx_r1.selectedFiles);
            i0.ɵɵattribute("aria-label", ctx_r1.uploadButtonText);
            i0.ɵɵadvance(1);
            i0.ɵɵtextInterpolate1(" ", ctx_r1.uploadButtonText, " ");
        }
    }
    function MatFileUploadComponent_button_10_mat_icon_1_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵelementStart(0, "mat-icon");
            i0.ɵɵtext(1, "close");
            i0.ɵɵelementEnd();
        }
    }
    function MatFileUploadComponent_button_10_mat_icon_2_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵelement(0, "mat-icon", 10);
        }
        if (rf & 2) {
            var ctx_r6 = i0.ɵɵnextContext(2);
            i0.ɵɵproperty("svgIcon", ctx_r6.customSvgIcon);
        }
    }
    function MatFileUploadComponent_button_10_Template(rf, ctx) {
        if (rf & 1) {
            var _r8_1 = i0.ɵɵgetCurrentView();
            i0.ɵɵelementStart(0, "button", 7);
            i0.ɵɵlistener("click", function MatFileUploadComponent_button_10_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r8_1); var ctx_r7 = i0.ɵɵnextContext(); return ctx_r7.filesChanged(null); });
            i0.ɵɵtemplate(1, MatFileUploadComponent_button_10_mat_icon_1_Template, 2, 0, "mat-icon", 8);
            i0.ɵɵtemplate(2, MatFileUploadComponent_button_10_mat_icon_2_Template, 1, 1, "mat-icon", 9);
            i0.ɵɵelementEnd();
        }
        if (rf & 2) {
            var ctx_r2 = i0.ɵɵnextContext();
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", !ctx_r2.customSvgIcon);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx_r2.customSvgIcon);
        }
    }
    var MatFileUploadComponent = /** @class */ (function () {
        function MatFileUploadComponent() {
            this.labelText = 'Select File(s)';
            this.selectButtonText = 'Select File(s)';
            this.selectFilesButtonType = 'button';
            this.uploadButtonText = 'Upload File(s)';
            this.uploadButtonType = 'button';
            this.allowMultipleFiles = false;
            this.showUploadButton = true;
            this.acceptedTypes = '*.*';
            this.customSvgIcon = null;
            this.uploadClicked = new i0.EventEmitter();
            this.selectedFilesChanged = new i0.EventEmitter();
            this.selectedFileText = '';
        }
        MatFileUploadComponent.prototype.filesChanged = function (files) {
            this.selectedFiles = files;
            this.selectedFilesChanged.emit(this.selectedFiles);
            if (this.selectedFiles) {
                var numSelectedFiles = this.selectedFiles.length;
                this.selectedFileText =
                    numSelectedFiles === 1
                        ? this.selectedFiles[0].name
                        : numSelectedFiles + " files selected";
            }
            else {
                this.selectedFileText = '';
                this.resetFileInput();
            }
        };
        MatFileUploadComponent.prototype.uploadFiles = function () {
            this.uploadClicked.emit(this.selectedFiles);
            this.resetFileInput();
        };
        MatFileUploadComponent.prototype.resetFileInput = function () {
            this.fileInputRef.nativeElement.value = '';
        };
        return MatFileUploadComponent;
    }());
    MatFileUploadComponent.ɵfac = function MatFileUploadComponent_Factory(t) { return new (t || MatFileUploadComponent)(); };
    MatFileUploadComponent.ɵcmp = i0.ɵɵdefineComponent({ type: MatFileUploadComponent, selectors: [["mat-file-upload"]], viewQuery: function MatFileUploadComponent_Query(rf, ctx) {
            if (rf & 1) {
                i0.ɵɵviewQuery(_c0, 1);
            }
            if (rf & 2) {
                var _t = void 0;
                i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.fileInputRef = _t.first);
            }
        }, inputs: { labelText: "labelText", selectButtonText: "selectButtonText", selectFilesButtonType: "selectFilesButtonType", uploadButtonText: "uploadButtonText", uploadButtonType: "uploadButtonType", allowMultipleFiles: "allowMultipleFiles", showUploadButton: "showUploadButton", acceptedTypes: "acceptedTypes", customSvgIcon: "customSvgIcon" }, outputs: { uploadClicked: "uploadClicked", selectedFilesChanged: "selectedFilesChanged" }, decls: 11, vars: 9, consts: [[1, "file-input-text"], ["mat-button", "", "color", "primary", 1, "file-input-button", 3, "type", "click"], ["type", "file", 2, "display", "none", 3, "accept", "multiple", "change"], ["fileInput", ""], ["mat-raised-button", "", "color", "primary", "class", "file-input-button", 3, "type", "disabled", "click", 4, "ngIf"], ["mat-icon-button", "", "type", "button", "aria-label", "Remove Selected File(s)", 3, "click", 4, "ngIf"], ["mat-raised-button", "", "color", "primary", 1, "file-input-button", 3, "type", "disabled", "click"], ["mat-icon-button", "", "type", "button", "aria-label", "Remove Selected File(s)", 3, "click"], [4, "ngIf"], [3, "svgIcon", 4, "ngIf"], [3, "svgIcon"]], template: function MatFileUploadComponent_Template(rf, ctx) {
            if (rf & 1) {
                var _r9_1 = i0.ɵɵgetCurrentView();
                i0.ɵɵelementStart(0, "span", 0);
                i0.ɵɵtext(1);
                i0.ɵɵelementEnd();
                i0.ɵɵelementStart(2, "button", 1);
                i0.ɵɵlistener("click", function MatFileUploadComponent_Template_button_click_2_listener() { i0.ɵɵrestoreView(_r9_1); var _r0 = i0.ɵɵreference(6); return _r0.click(); });
                i0.ɵɵelementStart(3, "span");
                i0.ɵɵtext(4);
                i0.ɵɵelementEnd();
                i0.ɵɵelementStart(5, "input", 2, 3);
                i0.ɵɵlistener("change", function MatFileUploadComponent_Template_input_change_5_listener($event) { return ctx.filesChanged($event.target.files); });
                i0.ɵɵelementEnd();
                i0.ɵɵelementEnd();
                i0.ɵɵtemplate(7, MatFileUploadComponent_button_7_Template, 2, 4, "button", 4);
                i0.ɵɵelementStart(8, "span", 0);
                i0.ɵɵtext(9);
                i0.ɵɵelementEnd();
                i0.ɵɵtemplate(10, MatFileUploadComponent_button_10_Template, 3, 2, "button", 5);
            }
            if (rf & 2) {
                i0.ɵɵadvance(1);
                i0.ɵɵtextInterpolate(ctx.labelText);
                i0.ɵɵadvance(1);
                i0.ɵɵproperty("type", ctx.selectFilesButtonType);
                i0.ɵɵattribute("aria-label", ctx.selectButtonText);
                i0.ɵɵadvance(2);
                i0.ɵɵtextInterpolate(ctx.selectButtonText);
                i0.ɵɵadvance(1);
                i0.ɵɵproperty("accept", ctx.acceptedTypes)("multiple", ctx.allowMultipleFiles);
                i0.ɵɵadvance(2);
                i0.ɵɵproperty("ngIf", ctx.showUploadButton);
                i0.ɵɵadvance(2);
                i0.ɵɵtextInterpolate(ctx.selectedFileText);
                i0.ɵɵadvance(1);
                i0.ɵɵproperty("ngIf", ctx.selectedFiles != null && ctx.selectedFiles.length > 0);
            }
        }, directives: [i1.MatButton, i2.NgIf, i3.MatIcon], styles: [".file-input-button[_ngcontent-%COMP%] { margin-right: 8px !important }", ".file-input-text[_ngcontent-%COMP%] { font-size: 14px !important; margin-right: 8px !important }"] });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(MatFileUploadComponent, [{
                type: i0.Component,
                args: [{
                        selector: 'mat-file-upload',
                        template: "\n    <span class=\"file-input-text\">{{ labelText }}</span>\n    <button\n      mat-button\n      [type]=\"selectFilesButtonType\"\n      color=\"primary\"\n      class=\"file-input-button\"\n      (click)=\"fileInput.click()\"\n      [attr.aria-label]=\"selectButtonText\"\n    >\n      <span>{{ selectButtonText }}</span>\n      <input\n        #fileInput\n        type=\"file\"\n        style=\"display: none\"\n        [accept]=\"acceptedTypes\"\n        [multiple]=\"allowMultipleFiles\"\n        (change)=\"filesChanged($event.target.files)\"\n      />\n    </button>\n    <button\n      mat-raised-button\n      [type]=\"uploadButtonType\"\n      color=\"primary\"\n      class=\"file-input-button\"\n      [disabled]=\"!selectedFiles\"\n      (click)=\"uploadFiles()\"\n      *ngIf=\"showUploadButton\"\n      [attr.aria-label]=\"uploadButtonText\"\n    >\n      {{ uploadButtonText }}\n    </button>\n    <span class=\"file-input-text\">{{ selectedFileText }}</span>\n    <button\n      mat-icon-button\n      (click)=\"filesChanged(null)\"\n      type=\"button\"\n      aria-label=\"Remove Selected File(s)\"\n       *ngIf=\"selectedFiles != null && selectedFiles.length > 0\"\n    >\n        <mat-icon *ngIf=\"!customSvgIcon\">close</mat-icon>\n        <mat-icon *ngIf=\"customSvgIcon\" [svgIcon]=\"customSvgIcon\"></mat-icon>\n    </button>\n  ",
                        styles: [
                            '.file-input-button { margin-right: 8px !important }',
                            '.file-input-text { font-size: 14px !important; margin-right: 8px !important }',
                        ],
                    }]
            }], null, { labelText: [{
                    type: i0.Input
                }], selectButtonText: [{
                    type: i0.Input
                }], selectFilesButtonType: [{
                    type: i0.Input
                }], uploadButtonText: [{
                    type: i0.Input
                }], uploadButtonType: [{
                    type: i0.Input
                }], allowMultipleFiles: [{
                    type: i0.Input
                }], showUploadButton: [{
                    type: i0.Input
                }], acceptedTypes: [{
                    type: i0.Input
                }], customSvgIcon: [{
                    type: i0.Input
                }], uploadClicked: [{
                    type: i0.Output
                }], selectedFilesChanged: [{
                    type: i0.Output
                }], fileInputRef: [{
                    type: i0.ViewChild,
                    args: ['fileInput']
                }] });
    })();

    var MatFileUploadModule = /** @class */ (function () {
        function MatFileUploadModule() {
        }
        return MatFileUploadModule;
    }());
    MatFileUploadModule.ɵmod = i0.ɵɵdefineNgModule({ type: MatFileUploadModule });
    MatFileUploadModule.ɵinj = i0.ɵɵdefineInjector({ factory: function MatFileUploadModule_Factory(t) { return new (t || MatFileUploadModule)(); }, imports: [[
                i2.CommonModule,
                i1.MatButtonModule,
                i3.MatIconModule,
                input.MatInputModule,
                select.MatSelectModule,
                forms.FormsModule,
            ]] });
    (function () {
        (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(MatFileUploadModule, { declarations: [MatFileUploadComponent], imports: [i2.CommonModule,
                i1.MatButtonModule,
                i3.MatIconModule,
                input.MatInputModule,
                select.MatSelectModule,
                forms.FormsModule], exports: [MatFileUploadComponent] });
    })();
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(MatFileUploadModule, [{
                type: i0.NgModule,
                args: [{
                        declarations: [MatFileUploadComponent],
                        imports: [
                            i2.CommonModule,
                            i1.MatButtonModule,
                            i3.MatIconModule,
                            input.MatInputModule,
                            select.MatSelectModule,
                            forms.FormsModule,
                        ],
                        exports: [MatFileUploadComponent],
                    }]
            }], null, null);
    })();

    /*
     * Public API Surface of mat-file-upload
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MatFileUploadComponent = MatFileUploadComponent;
    exports.MatFileUploadModule = MatFileUploadModule;
    exports.MatFileUploadService = MatFileUploadService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
