// jqwidgets-ng/bundles/jqwidgets-ng-jqxbulletchart.umd.js
require('../jqwidgets/jqxcore');
require('../jqwidgets/jqxdata');
require('../jqwidgets/jqxtooltip');
require('../jqwidgets/jqxbulletchart');
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('jqwidgets-ng/jqxbulletchart', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['jqwidgets-ng'] = global['jqwidgets-ng'] || {}, global['jqwidgets-ng'].jqxbulletchart = {}), global.ng.core));
}(this, (function (exports, core) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var jqxBulletChartComponent = /** @class */ (function () {
        function jqxBulletChartComponent(containerElement) {
            this.autoCreate = true;
            this.properties = ['animationDuration', 'barSize', 'description', 'disabled', 'height', 'labelsFormat', 'labelsFormatFunction', 'orientation', 'pointer', 'rtl', 'ranges', 'showTooltip', 'target', 'ticks', 'title', 'tooltipFormatFunction', 'width'];
            // jqxBulletChartComponent events
            this.onChange = new core.EventEmitter();
            this.elementRef = containerElement;
        }
        jqxBulletChartComponent.prototype.ngOnInit = function () {
            if (this.autoCreate) {
                this.createComponent();
            }
        };
        ;
        jqxBulletChartComponent.prototype.ngOnChanges = function (changes) {
            if (this.host) {
                for (var i = 0; i < this.properties.length; i++) {
                    var attrName = 'attr' + this.properties[i].substring(0, 1).toUpperCase() + this.properties[i].substring(1);
                    var areEqual = false;
                    if (this[attrName] !== undefined) {
                        if (typeof this[attrName] === 'object') {
                            if (this[attrName] instanceof Array) {
                                areEqual = this.arraysEqual(this[attrName], this.host.jqxBulletChart(this.properties[i]));
                            }
                            if (areEqual) {
                                return false;
                            }
                            this.host.jqxBulletChart(this.properties[i], this[attrName]);
                            continue;
                        }
                        if (this[attrName] !== this.host.jqxBulletChart(this.properties[i])) {
                            this.host.jqxBulletChart(this.properties[i], this[attrName]);
                        }
                    }
                }
            }
        };
        jqxBulletChartComponent.prototype.arraysEqual = function (attrValue, hostValue) {
            if ((attrValue && !hostValue) || (!attrValue && hostValue)) {
                return false;
            }
            if (attrValue.length != hostValue.length) {
                return false;
            }
            for (var i = 0; i < attrValue.length; i++) {
                if (attrValue[i] !== hostValue[i]) {
                    return false;
                }
            }
            return true;
        };
        jqxBulletChartComponent.prototype.manageAttributes = function () {
            var options = {};
            for (var i = 0; i < this.properties.length; i++) {
                var attrName = 'attr' + this.properties[i].substring(0, 1).toUpperCase() + this.properties[i].substring(1);
                if (this[attrName] !== undefined) {
                    options[this.properties[i]] = this[attrName];
                }
            }
            return options;
        };
        jqxBulletChartComponent.prototype.moveClasses = function (parentEl, childEl) {
            var _a;
            var classes = parentEl.classList;
            if (classes.length > 0) {
                (_a = childEl.classList).add.apply(_a, __spread(classes));
            }
            parentEl.className = '';
        };
        jqxBulletChartComponent.prototype.moveStyles = function (parentEl, childEl) {
            var style = parentEl.style.cssText;
            childEl.style.cssText = style;
            parentEl.style.cssText = '';
        };
        jqxBulletChartComponent.prototype.createComponent = function (options) {
            if (this.host) {
                return;
            }
            if (options) {
                JQXLite.extend(options, this.manageAttributes());
            }
            else {
                options = this.manageAttributes();
            }
            this.host = JQXLite(this.elementRef.nativeElement.firstChild);
            this.moveClasses(this.elementRef.nativeElement, this.host[0]);
            this.moveStyles(this.elementRef.nativeElement, this.host[0]);
            this.__wireEvents__();
            this.widgetObject = jqwidgets.createInstance(this.host, 'jqxBulletChart', options);
        };
        jqxBulletChartComponent.prototype.createWidget = function (options) {
            this.createComponent(options);
        };
        jqxBulletChartComponent.prototype.__updateRect__ = function () {
            if (this.host)
                this.host.css({ width: this.attrWidth, height: this.attrHeight });
        };
        jqxBulletChartComponent.prototype.setOptions = function (options) {
            this.host.jqxBulletChart('setOptions', options);
        };
        // jqxBulletChartComponent properties
        jqxBulletChartComponent.prototype.animationDuration = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('animationDuration', arg);
            }
            else {
                return this.host.jqxBulletChart('animationDuration');
            }
        };
        jqxBulletChartComponent.prototype.barSize = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('barSize', arg);
            }
            else {
                return this.host.jqxBulletChart('barSize');
            }
        };
        jqxBulletChartComponent.prototype.description = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('description', arg);
            }
            else {
                return this.host.jqxBulletChart('description');
            }
        };
        jqxBulletChartComponent.prototype.disabled = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('disabled', arg);
            }
            else {
                return this.host.jqxBulletChart('disabled');
            }
        };
        jqxBulletChartComponent.prototype.height = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('height', arg);
            }
            else {
                return this.host.jqxBulletChart('height');
            }
        };
        jqxBulletChartComponent.prototype.labelsFormat = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('labelsFormat', arg);
            }
            else {
                return this.host.jqxBulletChart('labelsFormat');
            }
        };
        jqxBulletChartComponent.prototype.labelsFormatFunction = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('labelsFormatFunction', arg);
            }
            else {
                return this.host.jqxBulletChart('labelsFormatFunction');
            }
        };
        jqxBulletChartComponent.prototype.orientation = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('orientation', arg);
            }
            else {
                return this.host.jqxBulletChart('orientation');
            }
        };
        jqxBulletChartComponent.prototype.pointer = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('pointer', arg);
            }
            else {
                return this.host.jqxBulletChart('pointer');
            }
        };
        jqxBulletChartComponent.prototype.rtl = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('rtl', arg);
            }
            else {
                return this.host.jqxBulletChart('rtl');
            }
        };
        jqxBulletChartComponent.prototype.ranges = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('ranges', arg);
            }
            else {
                return this.host.jqxBulletChart('ranges');
            }
        };
        jqxBulletChartComponent.prototype.showTooltip = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('showTooltip', arg);
            }
            else {
                return this.host.jqxBulletChart('showTooltip');
            }
        };
        jqxBulletChartComponent.prototype.target = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('target', arg);
            }
            else {
                return this.host.jqxBulletChart('target');
            }
        };
        jqxBulletChartComponent.prototype.ticks = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('ticks', arg);
            }
            else {
                return this.host.jqxBulletChart('ticks');
            }
        };
        jqxBulletChartComponent.prototype.title = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('title', arg);
            }
            else {
                return this.host.jqxBulletChart('title');
            }
        };
        jqxBulletChartComponent.prototype.tooltipFormatFunction = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('tooltipFormatFunction', arg);
            }
            else {
                return this.host.jqxBulletChart('tooltipFormatFunction');
            }
        };
        jqxBulletChartComponent.prototype.width = function (arg) {
            if (arg !== undefined) {
                this.host.jqxBulletChart('width', arg);
            }
            else {
                return this.host.jqxBulletChart('width');
            }
        };
        // jqxBulletChartComponent functions
        jqxBulletChartComponent.prototype.destroy = function () {
            this.host.jqxBulletChart('destroy');
        };
        jqxBulletChartComponent.prototype.render = function () {
            this.host.jqxBulletChart('render');
        };
        jqxBulletChartComponent.prototype.refresh = function () {
            this.host.jqxBulletChart('refresh');
        };
        jqxBulletChartComponent.prototype.val = function (value) {
            if (value !== undefined) {
                return this.host.jqxBulletChart('val', value);
            }
            else {
                return this.host.jqxBulletChart('val');
            }
        };
        ;
        jqxBulletChartComponent.prototype.__wireEvents__ = function () {
            var _this = this;
            this.host.on('change', function (eventData) { _this.onChange.emit(eventData); });
        };
        return jqxBulletChartComponent;
    }()); //jqxBulletChartComponent
    jqxBulletChartComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'jqxBulletChart',
                    template: '<div><ng-content></ng-content></div>'
                },] }
    ];
    jqxBulletChartComponent.ctorParameters = function () { return [
        { type: core.ElementRef }
    ]; };
    jqxBulletChartComponent.propDecorators = {
        attrAnimationDuration: [{ type: core.Input, args: ['animationDuration',] }],
        attrBarSize: [{ type: core.Input, args: ['barSize',] }],
        attrDescription: [{ type: core.Input, args: ['description',] }],
        attrDisabled: [{ type: core.Input, args: ['disabled',] }],
        attrLabelsFormat: [{ type: core.Input, args: ['labelsFormat',] }],
        attrLabelsFormatFunction: [{ type: core.Input, args: ['labelsFormatFunction',] }],
        attrOrientation: [{ type: core.Input, args: ['orientation',] }],
        attrPointer: [{ type: core.Input, args: ['pointer',] }],
        attrRtl: [{ type: core.Input, args: ['rtl',] }],
        attrRanges: [{ type: core.Input, args: ['ranges',] }],
        attrShowTooltip: [{ type: core.Input, args: ['showTooltip',] }],
        attrTarget: [{ type: core.Input, args: ['target',] }],
        attrTicks: [{ type: core.Input, args: ['ticks',] }],
        attrTitle: [{ type: core.Input, args: ['title',] }],
        attrTooltipFormatFunction: [{ type: core.Input, args: ['tooltipFormatFunction',] }],
        attrWidth: [{ type: core.Input, args: ['width',] }],
        attrHeight: [{ type: core.Input, args: ['height',] }],
        autoCreate: [{ type: core.Input, args: ['auto-create',] }],
        onChange: [{ type: core.Output }]
    };

    var jqxBulletChartModule = /** @class */ (function () {
        function jqxBulletChartModule() {
        }
        return jqxBulletChartModule;
    }());
    jqxBulletChartModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [],
                    declarations: [jqxBulletChartComponent],
                    exports: [jqxBulletChartComponent]
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.jqxBulletChartComponent = jqxBulletChartComponent;
    exports.jqxBulletChartModule = jqxBulletChartModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
