// jqwidgets-ng/bundles/jqwidgets-ng-jqxdropdownlist.umd.js
require('../jqwidgets/jqxcore');
require('../jqwidgets/jqxdata');
require('../jqwidgets/jqxbuttons');
require('../jqwidgets/jqxscrollbar');
require('../jqwidgets/jqxlistbox');
require('../jqwidgets/jqxdropdownlist');
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('jqwidgets-ng/jqxdropdownlist', ['exports', '@angular/core', '@angular/forms'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['jqwidgets-ng'] = global['jqwidgets-ng'] || {}, global['jqwidgets-ng'].jqxdropdownlist = {}), global.ng.core, global.ng.forms));
}(this, (function (exports, core, forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var noop = function () { };
    var ɵ0 = noop;
    var CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return jqxDropDownListComponent; }),
        multi: true
    };
    var jqxDropDownListComponent = /** @class */ (function () {
        function jqxDropDownListComponent(containerElement) {
            this.autoCreate = true;
            this.properties = ['autoOpen', 'autoDropDownHeight', 'animationType', 'checkboxes', 'closeDelay', 'disabled', 'displayMember', 'dropDownHorizontalAlignment', 'dropDownVerticalAlignment', 'dropDownHeight', 'dropDownWidth', 'enableSelection', 'enableBrowserBoundsDetection', 'enableHover', 'filterable', 'filterHeight', 'filterDelay', 'filterPlaceHolder', 'height', 'incrementalSearch', 'incrementalSearchDelay', 'itemHeight', 'openDelay', 'placeHolder', 'popupZIndex', 'rtl', 'renderer', 'selectionRenderer', 'searchMode', 'source', 'selectedIndex', 'theme', 'template', 'valueMember', 'width'];
            this.onTouchedCallback = noop;
            this.onChangeCallback = noop;
            // jqxDropDownListComponent events
            this.onBindingComplete = new core.EventEmitter();
            this.onClose = new core.EventEmitter();
            this.onCheckChange = new core.EventEmitter();
            this.onChange = new core.EventEmitter();
            this.onOpen = new core.EventEmitter();
            this.onSelect = new core.EventEmitter();
            this.onUnselect = new core.EventEmitter();
            this.elementRef = containerElement;
        }
        jqxDropDownListComponent.prototype.ngOnInit = function () {
        };
        ;
        jqxDropDownListComponent.prototype.ngAfterViewInit = function () {
            var children = JQXLite(this.elementRef.nativeElement.children).find('li');
            var html = '';
            var options = {};
            if (children.length > 0) {
                this.container = document.createElement('div');
                html = this.elementRef.nativeElement.innerHTML;
                this.container.appendChild(this.elementRef.nativeElement.firstChild);
                this.elementRef.nativeElement.innerHTML = html;
                this.content = html;
                var result = JQXLite.jqx.parseSourceTag(this.container);
                options['source'] = result.items;
            }
            if (this.autoCreate) {
                this.createComponent(options);
            }
        };
        ;
        jqxDropDownListComponent.prototype.ngAfterViewChecked = function () {
            if (this.container) {
                if (this.content !== this.container.innerHTML) {
                    this.content = this.container.innerHTML;
                    var result = JQXLite.jqx.parseSourceTag(this.container);
                    this.host.jqxDropDownList({ source: result.items });
                }
            }
        };
        ;
        jqxDropDownListComponent.prototype.ngOnChanges = function (changes) {
            if (this.host) {
                for (var i = 0; i < this.properties.length; i++) {
                    var attrName = 'attr' + this.properties[i].substring(0, 1).toUpperCase() + this.properties[i].substring(1);
                    var areEqual = false;
                    if (this[attrName] !== undefined) {
                        if (typeof this[attrName] === 'object') {
                            if (this[attrName] instanceof Array) {
                                areEqual = this.arraysEqual(this[attrName], this.host.jqxDropDownList(this.properties[i]));
                            }
                            if (areEqual) {
                                return false;
                            }
                            this.host.jqxDropDownList(this.properties[i], this[attrName]);
                            continue;
                        }
                        if (this[attrName] !== this.host.jqxDropDownList(this.properties[i])) {
                            this.host.jqxDropDownList(this.properties[i], this[attrName]);
                        }
                    }
                }
            }
        };
        jqxDropDownListComponent.prototype.arraysEqual = function (attrValue, hostValue) {
            if ((attrValue && !hostValue) || (!attrValue && hostValue)) {
                return false;
            }
            if (attrValue.length != hostValue.length) {
                return false;
            }
            for (var i = 0; i < attrValue.length; i++) {
                if (attrValue[i] !== hostValue[i]) {
                    return false;
                }
            }
            return true;
        };
        jqxDropDownListComponent.prototype.manageAttributes = function () {
            var options = {};
            for (var i = 0; i < this.properties.length; i++) {
                var attrName = 'attr' + this.properties[i].substring(0, 1).toUpperCase() + this.properties[i].substring(1);
                if (this[attrName] !== undefined) {
                    options[this.properties[i]] = this[attrName];
                }
            }
            return options;
        };
        jqxDropDownListComponent.prototype.moveClasses = function (parentEl, childEl) {
            var _a;
            var classes = parentEl.classList;
            if (classes.length > 0) {
                (_a = childEl.classList).add.apply(_a, __spread(classes));
            }
            parentEl.className = '';
        };
        jqxDropDownListComponent.prototype.moveStyles = function (parentEl, childEl) {
            var style = parentEl.style.cssText;
            childEl.style.cssText = style;
            parentEl.style.cssText = '';
        };
        jqxDropDownListComponent.prototype.createComponent = function (options) {
            if (this.host) {
                return;
            }
            if (options) {
                JQXLite.extend(options, this.manageAttributes());
            }
            else {
                options = this.manageAttributes();
            }
            this.host = JQXLite(this.elementRef.nativeElement.firstChild);
            this.moveClasses(this.elementRef.nativeElement, this.host[0]);
            this.moveStyles(this.elementRef.nativeElement, this.host[0]);
            this.__wireEvents__();
            this.widgetObject = jqwidgets.createInstance(this.host, 'jqxDropDownList', options);
        };
        jqxDropDownListComponent.prototype.createWidget = function (options) {
            this.createComponent(options);
        };
        jqxDropDownListComponent.prototype.__updateRect__ = function () {
            if (this.host)
                this.host.css({ width: this.attrWidth, height: this.attrHeight });
        };
        jqxDropDownListComponent.prototype.writeValue = function (value) {
            if (this.widgetObject) {
                this.onChangeCallback(this.host.val());
                this.widgetObject.clearSelection();
                if (value) {
                    this.widgetObject.selectItem(this.widgetObject.getItemByValue(value));
                }
            }
            if (this.host && (value === null || value === undefined)) {
                this.host.jqxDropDownList('val', '');
            }
        };
        jqxDropDownListComponent.prototype.registerOnChange = function (fn) {
            this.onChangeCallback = fn;
        };
        jqxDropDownListComponent.prototype.registerOnTouched = function (fn) {
            this.onTouchedCallback = fn;
        };
        jqxDropDownListComponent.prototype.setOptions = function (options) {
            this.host.jqxDropDownList('setOptions', options);
        };
        // jqxDropDownListComponent properties
        jqxDropDownListComponent.prototype.autoOpen = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('autoOpen', arg);
            }
            else {
                return this.host.jqxDropDownList('autoOpen');
            }
        };
        jqxDropDownListComponent.prototype.autoDropDownHeight = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('autoDropDownHeight', arg);
            }
            else {
                return this.host.jqxDropDownList('autoDropDownHeight');
            }
        };
        jqxDropDownListComponent.prototype.animationType = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('animationType', arg);
            }
            else {
                return this.host.jqxDropDownList('animationType');
            }
        };
        jqxDropDownListComponent.prototype.checkboxes = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('checkboxes', arg);
            }
            else {
                return this.host.jqxDropDownList('checkboxes');
            }
        };
        jqxDropDownListComponent.prototype.closeDelay = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('closeDelay', arg);
            }
            else {
                return this.host.jqxDropDownList('closeDelay');
            }
        };
        jqxDropDownListComponent.prototype.disabled = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('disabled', arg);
            }
            else {
                return this.host.jqxDropDownList('disabled');
            }
        };
        jqxDropDownListComponent.prototype.displayMember = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('displayMember', arg);
            }
            else {
                return this.host.jqxDropDownList('displayMember');
            }
        };
        jqxDropDownListComponent.prototype.dropDownHorizontalAlignment = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('dropDownHorizontalAlignment', arg);
            }
            else {
                return this.host.jqxDropDownList('dropDownHorizontalAlignment');
            }
        };
        jqxDropDownListComponent.prototype.dropDownVerticalAlignment = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('dropDownVerticalAlignment', arg);
            }
            else {
                return this.host.jqxDropDownList('dropDownVerticalAlignment');
            }
        };
        jqxDropDownListComponent.prototype.dropDownHeight = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('dropDownHeight', arg);
            }
            else {
                return this.host.jqxDropDownList('dropDownHeight');
            }
        };
        jqxDropDownListComponent.prototype.dropDownWidth = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('dropDownWidth', arg);
            }
            else {
                return this.host.jqxDropDownList('dropDownWidth');
            }
        };
        jqxDropDownListComponent.prototype.enableSelection = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('enableSelection', arg);
            }
            else {
                return this.host.jqxDropDownList('enableSelection');
            }
        };
        jqxDropDownListComponent.prototype.enableBrowserBoundsDetection = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('enableBrowserBoundsDetection', arg);
            }
            else {
                return this.host.jqxDropDownList('enableBrowserBoundsDetection');
            }
        };
        jqxDropDownListComponent.prototype.enableHover = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('enableHover', arg);
            }
            else {
                return this.host.jqxDropDownList('enableHover');
            }
        };
        jqxDropDownListComponent.prototype.filterable = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('filterable', arg);
            }
            else {
                return this.host.jqxDropDownList('filterable');
            }
        };
        jqxDropDownListComponent.prototype.filterHeight = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('filterHeight', arg);
            }
            else {
                return this.host.jqxDropDownList('filterHeight');
            }
        };
        jqxDropDownListComponent.prototype.filterDelay = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('filterDelay', arg);
            }
            else {
                return this.host.jqxDropDownList('filterDelay');
            }
        };
        jqxDropDownListComponent.prototype.filterPlaceHolder = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('filterPlaceHolder', arg);
            }
            else {
                return this.host.jqxDropDownList('filterPlaceHolder');
            }
        };
        jqxDropDownListComponent.prototype.height = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('height', arg);
            }
            else {
                return this.host.jqxDropDownList('height');
            }
        };
        jqxDropDownListComponent.prototype.incrementalSearch = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('incrementalSearch', arg);
            }
            else {
                return this.host.jqxDropDownList('incrementalSearch');
            }
        };
        jqxDropDownListComponent.prototype.incrementalSearchDelay = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('incrementalSearchDelay', arg);
            }
            else {
                return this.host.jqxDropDownList('incrementalSearchDelay');
            }
        };
        jqxDropDownListComponent.prototype.itemHeight = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('itemHeight', arg);
            }
            else {
                return this.host.jqxDropDownList('itemHeight');
            }
        };
        jqxDropDownListComponent.prototype.openDelay = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('openDelay', arg);
            }
            else {
                return this.host.jqxDropDownList('openDelay');
            }
        };
        jqxDropDownListComponent.prototype.placeHolder = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('placeHolder', arg);
            }
            else {
                return this.host.jqxDropDownList('placeHolder');
            }
        };
        jqxDropDownListComponent.prototype.popupZIndex = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('popupZIndex', arg);
            }
            else {
                return this.host.jqxDropDownList('popupZIndex');
            }
        };
        jqxDropDownListComponent.prototype.rtl = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('rtl', arg);
            }
            else {
                return this.host.jqxDropDownList('rtl');
            }
        };
        jqxDropDownListComponent.prototype.renderer = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('renderer', arg);
            }
            else {
                return this.host.jqxDropDownList('renderer');
            }
        };
        jqxDropDownListComponent.prototype.selectionRenderer = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('selectionRenderer', arg);
            }
            else {
                return this.host.jqxDropDownList('selectionRenderer');
            }
        };
        jqxDropDownListComponent.prototype.searchMode = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('searchMode', arg);
            }
            else {
                return this.host.jqxDropDownList('searchMode');
            }
        };
        jqxDropDownListComponent.prototype.source = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('source', arg);
            }
            else {
                return this.host.jqxDropDownList('source');
            }
        };
        jqxDropDownListComponent.prototype.selectedIndex = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('selectedIndex', arg);
            }
            else {
                return this.host.jqxDropDownList('selectedIndex');
            }
        };
        jqxDropDownListComponent.prototype.theme = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('theme', arg);
            }
            else {
                return this.host.jqxDropDownList('theme');
            }
        };
        jqxDropDownListComponent.prototype.template = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('template', arg);
            }
            else {
                return this.host.jqxDropDownList('template');
            }
        };
        jqxDropDownListComponent.prototype.valueMember = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('valueMember', arg);
            }
            else {
                return this.host.jqxDropDownList('valueMember');
            }
        };
        jqxDropDownListComponent.prototype.width = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            if (arg !== undefined) {
                this.host.jqxDropDownList('width', arg);
            }
            else {
                return this.host.jqxDropDownList('width');
            }
        };
        // jqxDropDownListComponent functions
        jqxDropDownListComponent.prototype.addItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('addItem', item);
        };
        jqxDropDownListComponent.prototype.clearSelection = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('clearSelection');
        };
        jqxDropDownListComponent.prototype.clear = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('clear');
        };
        jqxDropDownListComponent.prototype.close = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('close');
        };
        jqxDropDownListComponent.prototype.checkIndex = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('checkIndex', index);
        };
        jqxDropDownListComponent.prototype.checkItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('checkItem', item);
        };
        jqxDropDownListComponent.prototype.checkAll = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('checkAll');
        };
        jqxDropDownListComponent.prototype.clearFilter = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('clearFilter');
        };
        jqxDropDownListComponent.prototype.destroy = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('destroy');
        };
        jqxDropDownListComponent.prototype.disableItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('disableItem', item);
        };
        jqxDropDownListComponent.prototype.disableAt = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('disableAt', index);
        };
        jqxDropDownListComponent.prototype.enableItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('enableItem', item);
        };
        jqxDropDownListComponent.prototype.enableAt = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('enableAt', index);
        };
        jqxDropDownListComponent.prototype.ensureVisible = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('ensureVisible', index);
        };
        jqxDropDownListComponent.prototype.focus = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('focus');
        };
        jqxDropDownListComponent.prototype.getItem = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('getItem', index);
        };
        jqxDropDownListComponent.prototype.getItemByValue = function (itemValue) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('getItemByValue', itemValue);
        };
        jqxDropDownListComponent.prototype.getItems = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('getItems');
        };
        jqxDropDownListComponent.prototype.getCheckedItems = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('getCheckedItems');
        };
        jqxDropDownListComponent.prototype.getSelectedItem = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('getSelectedItem');
        };
        jqxDropDownListComponent.prototype.getSelectedIndex = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('getSelectedIndex');
        };
        jqxDropDownListComponent.prototype.insertAt = function (item, index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('insertAt', item, index);
        };
        jqxDropDownListComponent.prototype.isOpened = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            return this.host.jqxDropDownList('isOpened');
        };
        jqxDropDownListComponent.prototype.indeterminateIndex = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('indeterminateIndex', index);
        };
        jqxDropDownListComponent.prototype.indeterminateItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('indeterminateItem', item);
        };
        jqxDropDownListComponent.prototype.loadFromSelect = function (arg) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('loadFromSelect', arg);
        };
        jqxDropDownListComponent.prototype.open = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('open');
        };
        jqxDropDownListComponent.prototype.removeItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('removeItem', item);
        };
        jqxDropDownListComponent.prototype.removeAt = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('removeAt', index);
        };
        jqxDropDownListComponent.prototype.selectIndex = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('selectIndex', index);
        };
        jqxDropDownListComponent.prototype.selectItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('selectItem', item);
        };
        jqxDropDownListComponent.prototype.setContent = function (content) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('setContent', content);
        };
        jqxDropDownListComponent.prototype.updateItem = function (newItem, item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('updateItem', newItem, item);
        };
        jqxDropDownListComponent.prototype.updateAt = function (item, index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('updateAt', item, index);
        };
        jqxDropDownListComponent.prototype.unselectIndex = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('unselectIndex', index);
        };
        jqxDropDownListComponent.prototype.unselectItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('unselectItem', item);
        };
        jqxDropDownListComponent.prototype.uncheckIndex = function (index) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('uncheckIndex', index);
        };
        jqxDropDownListComponent.prototype.uncheckItem = function (item) {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('uncheckItem', item);
        };
        jqxDropDownListComponent.prototype.uncheckAll = function () {
            if (this.autoCreate && !this.host) {
                this.createComponent();
            }
            this.host.jqxDropDownList('uncheckAll');
        };
        jqxDropDownListComponent.prototype.val = function (value) {
            if (value !== undefined) {
                return this.host.jqxDropDownList('val', value);
            }
            else {
                return this.host.jqxDropDownList('val');
            }
        };
        ;
        jqxDropDownListComponent.prototype.__wireEvents__ = function () {
            var _this = this;
            this.host.on('bindingComplete', function (eventData) { _this.onBindingComplete.emit(eventData); });
            this.host.on('close', function (eventData) { _this.onClose.emit(eventData); });
            this.host.on('checkChange', function (eventData) { _this.onCheckChange.emit(eventData); });
            this.host.on('change', function (eventData) {
                _this.onChange.emit(eventData);
                if (eventData.args)
                    _this.onChangeCallback(eventData.args.item.value);
            });
            this.host.on('open', function (eventData) { _this.onOpen.emit(eventData); });
            this.host.on('select', function (eventData) { _this.onSelect.emit(eventData); });
            this.host.on('unselect', function (eventData) { _this.onUnselect.emit(eventData); });
        };
        return jqxDropDownListComponent;
    }()); //jqxDropDownListComponent
    jqxDropDownListComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'jqxDropDownList',
                    template: '<div><ng-content></ng-content></div>',
                    providers: [CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR],
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                },] }
    ];
    jqxDropDownListComponent.ctorParameters = function () { return [
        { type: core.ElementRef }
    ]; };
    jqxDropDownListComponent.propDecorators = {
        attrAutoOpen: [{ type: core.Input, args: ['autoOpen',] }],
        attrAutoDropDownHeight: [{ type: core.Input, args: ['autoDropDownHeight',] }],
        attrAnimationType: [{ type: core.Input, args: ['animationType',] }],
        attrCheckboxes: [{ type: core.Input, args: ['checkboxes',] }],
        attrCloseDelay: [{ type: core.Input, args: ['closeDelay',] }],
        attrDisabled: [{ type: core.Input, args: ['disabled',] }],
        attrDisplayMember: [{ type: core.Input, args: ['displayMember',] }],
        attrDropDownHorizontalAlignment: [{ type: core.Input, args: ['dropDownHorizontalAlignment',] }],
        attrDropDownVerticalAlignment: [{ type: core.Input, args: ['dropDownVerticalAlignment',] }],
        attrDropDownHeight: [{ type: core.Input, args: ['dropDownHeight',] }],
        attrDropDownWidth: [{ type: core.Input, args: ['dropDownWidth',] }],
        attrEnableSelection: [{ type: core.Input, args: ['enableSelection',] }],
        attrEnableBrowserBoundsDetection: [{ type: core.Input, args: ['enableBrowserBoundsDetection',] }],
        attrEnableHover: [{ type: core.Input, args: ['enableHover',] }],
        attrFilterable: [{ type: core.Input, args: ['filterable',] }],
        attrFilterHeight: [{ type: core.Input, args: ['filterHeight',] }],
        attrFilterDelay: [{ type: core.Input, args: ['filterDelay',] }],
        attrFilterPlaceHolder: [{ type: core.Input, args: ['filterPlaceHolder',] }],
        attrIncrementalSearch: [{ type: core.Input, args: ['incrementalSearch',] }],
        attrIncrementalSearchDelay: [{ type: core.Input, args: ['incrementalSearchDelay',] }],
        attrItemHeight: [{ type: core.Input, args: ['itemHeight',] }],
        attrOpenDelay: [{ type: core.Input, args: ['openDelay',] }],
        attrPlaceHolder: [{ type: core.Input, args: ['placeHolder',] }],
        attrPopupZIndex: [{ type: core.Input, args: ['popupZIndex',] }],
        attrRtl: [{ type: core.Input, args: ['rtl',] }],
        attrRenderer: [{ type: core.Input, args: ['renderer',] }],
        attrSelectionRenderer: [{ type: core.Input, args: ['selectionRenderer',] }],
        attrSearchMode: [{ type: core.Input, args: ['searchMode',] }],
        attrSource: [{ type: core.Input, args: ['source',] }],
        attrSelectedIndex: [{ type: core.Input, args: ['selectedIndex',] }],
        attrTheme: [{ type: core.Input, args: ['theme',] }],
        attrTemplate: [{ type: core.Input, args: ['template',] }],
        attrValueMember: [{ type: core.Input, args: ['valueMember',] }],
        attrWidth: [{ type: core.Input, args: ['width',] }],
        attrHeight: [{ type: core.Input, args: ['height',] }],
        autoCreate: [{ type: core.Input, args: ['auto-create',] }],
        onBindingComplete: [{ type: core.Output }],
        onClose: [{ type: core.Output }],
        onCheckChange: [{ type: core.Output }],
        onChange: [{ type: core.Output }],
        onOpen: [{ type: core.Output }],
        onSelect: [{ type: core.Output }],
        onUnselect: [{ type: core.Output }]
    };

    var jqxDropDownListModule = /** @class */ (function () {
        function jqxDropDownListModule() {
        }
        return jqxDropDownListModule;
    }());
    jqxDropDownListModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        forms.FormsModule
                    ],
                    declarations: [jqxDropDownListComponent],
                    exports: [jqxDropDownListComponent]
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.jqxDropDownListComponent = jqxDropDownListComponent;
    exports.jqxDropDownListModule = jqxDropDownListModule;
    exports.ɵ0 = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
